[SVG] Add approximate stroke-bounding-box computation for repainting

Constellation opened the pull request:
#### f7f897982b66445d96622a5b67145280e2ba2d81
<pre>
[SVG] Add approximate stroke-bounding-box computation for repainting
<a href="https://bugs.webkit.org/show_bug.cgi?id=263184">https://bugs.webkit.org/show_bug.cgi?id=263184</a>
rdar://116999242

Reviewed by Cameron McCormack.

This is one step of the patch series implementing approximate repainting rect for SVG path, derived from blink&apos;s work[1].

This patch adds approximate stroke-bounding-box (basically repaint bounding box) implementation. The core concept is that, for very compute costly element (like, path),
for repaint bounding box, we just report approximate repaint bounding box. This can be computed from stroke width etc. Like, let&apos;s
assume that stroke exists at the edge of fill bounding box. Then we can compute approximate repaint bounding box which is always larger-or-equal
to stroke bounding box.

Important thing in this patch is that we are not using approximate stroke-bounding-box computation yet. We will switch
it in the next patch with a few lines of changes. This patch is adding the implementation separately from enablement.

One of the hard thing is that our current implementation is relying on the computation order of this in the tree because of existence of
recursive / circular references of the repaint bounding box via markers. For example, LayoutTests/svg/custom/circular-marker-reference-4.svg
shows how markers can refer to the other marker in a circular manner. And in this case, our current implementation is just using stale
repaint bounding box for one of them, so probably the implementation is wrong already. But in this patch, we do not want to change the behavior.
As a result, we workaround the issue by eagerly compute stroke bounding box when markers exist in LegacyRenderSVGPath. This does not become
a problem in accuracy of stroke bounding box in LBSE since it is implementing SVG2, and SVG2 stroke bounding box does not include markers.

[1]: <a href="https://bugs.chromium.org/p/chromium/issues/detail?id=435097">https://bugs.chromium.org/p/chromium/issues/detail?id=435097</a>

* Source/WebCore/rendering/svg/RenderSVGContainer.cpp:
(WebCore::RenderSVGContainer::strokeBoundingBox const):
* Source/WebCore/rendering/svg/RenderSVGEllipse.cpp:
(WebCore::RenderSVGEllipse::updateShapeFromElement):
* Source/WebCore/rendering/svg/RenderSVGPath.cpp:
(WebCore::RenderSVGPath::updateShapeFromElement):
* Source/WebCore/rendering/svg/RenderSVGRect.cpp:
(WebCore::RenderSVGRect::updateShapeFromElement):
* Source/WebCore/rendering/svg/RenderSVGRoot.cpp:
(WebCore::RenderSVGRoot::strokeBoundingBox const):
* Source/WebCore/rendering/svg/RenderSVGShape.cpp:
(WebCore::RenderSVGShape::strokeBoundingBox const):
(WebCore::RenderSVGShape::approximateStrokeBoundingBox const):
(WebCore::RenderSVGShape::calculateApproximateScalingStrokeBoundingBox const):
(WebCore::RenderSVGShape::calculateApproximateNonScalingStrokeBoundingBox const):
(WebCore::RenderSVGShape::calculateApproximateStrokeBoundingBox const):
* Source/WebCore/rendering/svg/RenderSVGShape.h:
* Source/WebCore/rendering/svg/SVGBoundingBoxComputation.cpp:
(WebCore::SVGBoundingBoxComputation::handleShapeOrTextOrInline const):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGContainer.cpp:
(WebCore::LegacyRenderSVGContainer::updateCachedBoundaries):
(WebCore::LegacyRenderSVGContainer::strokeBoundingBox const):
(WebCore::LegacyRenderSVGContainer::repaintRectInLocalCoordinates const):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGContainer.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGPath.cpp:
(WebCore::LegacyRenderSVGPath::updateShapeFromElement):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRoot.cpp:
(WebCore::LegacyRenderSVGRoot::updateCachedBoundaries):
(WebCore::LegacyRenderSVGRoot::strokeBoundingBox const):
(WebCore::LegacyRenderSVGRoot::repaintRectInLocalCoordinates const):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRoot.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.cpp:
(WebCore::LegacyRenderSVGShape::strokeBoundingBox const):
(WebCore::LegacyRenderSVGShape::calculateApproximateScalingStrokeBoundingBox const):
(WebCore::LegacyRenderSVGShape::calculateApproximateNonScalingStrokeBoundingBox const):
(WebCore::LegacyRenderSVGShape::calculateApproximateStrokeBoundingBox const):
(WebCore::LegacyRenderSVGShape::updateRepaintBoundingBox):
(WebCore::LegacyRenderSVGShape::repaintRectInLocalCoordinates const):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.h:

Canonical link: <a href="https://commits.webkit.org/269492@main">https://commits.webkit.org/269492@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5638851c2f7d359e7696c6ca118be0895619dcac

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22757 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/420 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23844 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24665 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21063 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/23014 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/1493 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/16/builds/23283 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/21977 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22997 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/1493 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/19740 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25518 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/1493 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/20619 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/26832 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/1493 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/20862 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/24686 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/300 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/16/builds/23283 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/234 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/359 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2863 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/297 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebCore/rendering/svg/RenderSVGContainer.cpp:
```diff
@@ -112,6 +112,7 @@ void RenderSVGContainer::layoutChildren()
 FloatRect RenderSVGContainer::strokeBoundingBox() const
 {
     if (!m_strokeBoundingBox) {
+        m_strokeBoundingBox = FloatRect { };
         SVGBoundingBoxComputation boundingBoxComputation(*this);
         m_strokeBoundingBox = boundingBoxComputation.computeDecoratedBoundingBox(SVGBoundingBoxComputation::strokeBoundingBoxDecoration);
```
Assigning to `m_strokeBoundingBox` twice looks strange. Is this deliberate? Similar in other files.

Constellation replied:
Yes, this is deliberate. The reason is that our current LSBE code is not detecting recursion of this invoked from repaint bounding box computation for markers / SVG resources (like, filters, clippers, maskers). As a result, without filling it to empty stroke-bounding-box before starting computation, we will have infinite recursion in some of tests.

The example test is https://github.com/WebKit/WebKit/blob/main/LayoutTests/svg/custom/svg-reference-cycle-between-two-svg-elements.html. This is also referencing each other in a circular manner.

Right now, I cannot find the right behavior in that case from the spec. It looks like the spec does infinite recursion. Here, I assigned empty one before starting recursion so that we can stop the recursion since `m_strokeBoundingBox` is no longer `std::nullopt`. (So we do not need to crash).
Ideally we would like to have some specified behavior for that, but right now, it is just *undefined*. (Current our implementation would return the stale previous stroke-bounding-box)

The part of the issue of this is that we are using accurate painting bounding box for markers, which invokes stroke-bounding-box. So, marker -> accurate repaint bounding box -> stroke bounding box -> another marker's root's stroke bounding box -> another marker -> accurate repaint bounding box -> going to the first marker. Previously we were using non accurate repaint bounding box. Then in that case, we were using stale repaint bounding box. Either behavior looks incorrect anyway, and even spec does not answer to the right expected behavior (And it makes sense since this is repaint bounding box related issue, but I think we don't have a reasonable way to represent this recursive SVG). Thus, we are just keeping the current status-quo roughly: stopping the recursion.

heycam replied:
OK, can you please add a one line comment just above the first assignment, in all of the functions that do this, to explain why? Just something like:

```
// Initialize m_strokeBoundingBox before calling computeDecoratedBoundingBox, since recursively referenced markers can cause us to re-enter here.
```

Constellation replied:
Sounds good! Added.

heycam commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.cpp:
```diff
@@ -383,12 +385,107 @@ FloatRect LegacyRenderSVGShape::calculateStrokeBoundingBox() const
+    case ShapeType::Ellipse:
+    case ShapeType::Circle: {
+        break;
+    }
```
Nit: no braces.

Constellation replied:
Fixed!

heycam commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.cpp:
```diff
@@ -383,12 +385,107 @@ FloatRect LegacyRenderSVGShape::calculateStrokeBoundingBox() const
+    float delta = strokeWidth / 2;
+    switch (m_shapeType) {
+    case ShapeType::Empty: {
+        // // Spec: "A negative value is illegal. A value of zero disables rendering of the element."
```
Nit: two "//"s.

Constellation replied:
Fixed!

heycam commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.cpp:
```diff
@@ -383,12 +385,107 @@ FloatRect LegacyRenderSVGShape::calculateStrokeBoundingBox() const
 }
 
+FloatRect LegacyRenderSVGShape::calculateApproximateScalingStrokeBoundingBox(FloatRect fillBoundingBox) const
+{
```
To parallel the non-scaling one, let's:

```
ASSERT(style().svgStyle().hasStroke());
ASSERT(!hasNonScalingStroke());
```

Constellation replied:
Looks good! Changed.

heycam commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.cpp:
```diff
@@ -383,12 +385,107 @@ FloatRect LegacyRenderSVGShape::calculateStrokeBoundingBox() const
+    AffineTransform nonScalingTransform = nonScalingStrokeTransform();
+    if (std::optional<AffineTransform> inverse = nonScalingTransform.inverse()) {
+        Path* usePath = nonScalingStrokePath(m_path.get(), nonScalingTransform);
+        FloatRect strokeBoundingRect = calculateApproximateScalingStrokeBoundingBox(usePath->fastBoundingRect());
```
General comment, which I should've mentioned on the previous patches too: prefer to use `auto` / `auto*` / `auto&` unless it's confusing or unclear not to know the exact type or you explicitly want to use a type different from what `auto` would give you. Here I would use `auto` for all of these declarations. (Existing code might not already follow this guideline well.)

Constellation replied:
Thanks! I've changed :D

heycam commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.cpp:
```diff
@@ -383,12 +385,107 @@ FloatRect LegacyRenderSVGShape::calculateStrokeBoundingBox() const
+            approximateStrokeBoundingBox = calculateApproximateNonScalingStrokeBoundingBox(m_fillBoundingBox);
+        else
+            approximateStrokeBoundingBox = calculateApproximateScalingStrokeBoundingBox(m_fillBoundingBox);
+    }
```
Nit: to avoid some indentation I'd write it as:

```
if (!style().svgStyle().hasStroke())
    ...
else if (hasNonScalingStroke())
    ...
else
    ...
```

Constellation replied:
Sounds good! Changed.

heycam commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGPath.cpp:
```diff
@@ -63,6 +63,17 @@ void LegacyRenderSVGPath::updateShapeFromElement()
         m_shapeType = ShapeType::Path;
+
+    // FIXME: This should not exist. However, currently SVG is relying on ordering of calculation of strokeBoundingBox via layout() function
+    // for recursive SVGs (markers are pointing to each other recursively). If we move to SVG2 compuation, we no longer need this since SVG2 strokeBoundingBox
```
"computation"

Constellation replied:
Thanks, changed!

heycam commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGPath.cpp:
```diff
@@ -63,6 +63,17 @@ void LegacyRenderSVGPath::updateShapeFromElement()
+
+    // FIXME: This should not exist. However, currently SVG is relying on ordering of calculation of strokeBoundingBox via layout() function
+    // for recursive SVGs (markers are pointing to each other recursively). If we move to SVG2 compuation, we no longer need this since SVG2 strokeBoundingBox
+    // does not include markers rect (so we do not need to have this in LSBE). Right now, this exists only for RepaintRectCalculation::Accurate, and it should be removed once
```
This and the commit message makes me slightly confused, because (per my comment on the top level PR) "stroke bounding box" is not a concept that exists in SVG 1. The fact that our LegacyRenderSVG* classes talk about stroke bounding boxes is really just a poor name for repaint rects, I think.

Constellation replied:
OK, I mentioned to SVG2 stroke-bounding box here for clarification, and yeah this is unfortunate that Legacy and new SVG implementations are using stroke-bounding-box for different meanings...

heycam commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGPath.cpp:
```diff
@@ -63,6 +63,17 @@ void LegacyRenderSVGPath::updateShapeFromElement()
+    // FIXME: This should not exist. However, currently SVG is relying on ordering of calculation of strokeBoundingBox via layout() function
+    // for recursive SVGs (markers are pointing to each other recursively). If we move to SVG2 compuation, we no longer need this since SVG2 strokeBoundingBox
+    // does not include markers rect (so we do not need to have this in LSBE). Right now, this exists only for RepaintRectCalculation::Accurate, and it should be removed once
+    // 1. We fix checkInsersion / checkEnclosure implementations. Currently they are not aligned to what the spec requires.
```
"checkIntersection"

Constellation replied:
Fixed!

heycam commented on Source/WebCore/rendering/svg/RenderSVGShape.cpp:
```diff
@@ -388,6 +390,99 @@ FloatRect RenderSVGShape::calculateStrokeBoundingBox() const
+            approximateStrokeBoundingBox = calculateApproximateScalingStrokeBoundingBox(m_fillBoundingBox);
+    }
+
+    return adjustStrokeBoundingBoxForMarkersAndZeroLengthLinecaps(RepaintRectCalculation::Fast, approximateStrokeBoundingBox);
```
`RenderSVGShape::adjustStrokeBoundingBoxForMarkersAndZeroLengthLinecaps` should be renamed not to mention markers.

Constellation replied:
Sounds good, changed!

heycam commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.h:
```diff
@@ -123,11 +125,14 @@ class LegacyRenderSVGShape : public LegacyRenderSVGModelObject {
+
+    FloatRect calculateApproximateStrokeBoundingBox() const;
+    FloatRect calculateApproximateScalingStrokeBoundingBox(FloatRect fillBoundingBox) const;
+    FloatRect calculateApproximateNonScalingStrokeBoundingBox(FloatRect fillBoundingBox) const;
```
Please consider putting these two functions  (and maybe some of the guts of `calculateApproximateStrokeBoundingBox`?) in `SVGRenderSupport`, which is where we have some other code that is common to LBSE and Legacy.

Constellation replied:
Moved!

Constellation reviewed (commented):
I totally agree that accurate repaint bounding box should be removed. This is introduced just to keep the test results unchanged. But requiring accuracy for repaint bounding box is breaking the concept of that, and it is revealing the code places which is mixing repaint bounding box and other specified bounding boxes. Current accurate repaint bounding box is neither approximate repaint bounding box nor stroke bounding box. It exists only for DumpRenderTree and `checkEnclosure` / `checkInsertion` SVG functions, where we should not use repaint bounding box.

Constellation reviewed (commented):
Thanks!

Constellation committed:
[SVG] Add approximate stroke-bounding-box computation for repainting
https://bugs.webkit.org/show_bug.cgi?id=263184
rdar://116999242

Reviewed by Cameron McCormack.

This is one step of the patch series implementing approximate repainting rect for SVG path, derived from blink's work[1].

This patch adds approximate stroke-bounding-box (basically repaint bounding box) implementation. The core concept is that, for very compute costly element (like, path),
for repaint bounding box, we just report approximate repaint bounding box. This can be computed from stroke width etc. Like, let's
assume that stroke exists at the edge of fill bounding box. Then we can compute approximate repaint bounding box which is always larger-or-equal
to stroke bounding box.

Important thing in this patch is that we are not using approximate stroke-bounding-box computation yet. We will switch
it in the next patch with a few lines of changes. This patch is adding the implementation separately from enablement.

One of the hard thing is that our current implementation is relying on the computation order of this in the tree because of existence of
recursive / circular references of the repaint bounding box via markers. For example, LayoutTests/svg/custom/circular-marker-reference-4.svg
shows how markers can refer to the other marker in a circular manner. And in this case, our current implementation is just using stale
repaint bounding box for one of them, so probably the implementation is wrong already. But in this patch, we do not want to change the behavior.
As a result, we workaround the issue by eagerly compute stroke bounding box when markers exist in LegacyRenderSVGPath. This does not become
a problem in accuracy of stroke bounding box in LBSE since it is implementing SVG2, and SVG2 stroke bounding box does not include markers.

[1]: https://bugs.chromium.org/p/chromium/issues/detail?id=435097

* Source/WebCore/rendering/svg/RenderSVGContainer.cpp:
(WebCore::RenderSVGContainer::strokeBoundingBox const):
* Source/WebCore/rendering/svg/RenderSVGEllipse.cpp:
(WebCore::RenderSVGEllipse::updateShapeFromElement):
* Source/WebCore/rendering/svg/RenderSVGPath.cpp:
(WebCore::RenderSVGPath::updateShapeFromElement):
* Source/WebCore/rendering/svg/RenderSVGRect.cpp:
(WebCore::RenderSVGRect::updateShapeFromElement):
* Source/WebCore/rendering/svg/RenderSVGRoot.cpp:
(WebCore::RenderSVGRoot::strokeBoundingBox const):
* Source/WebCore/rendering/svg/RenderSVGShape.cpp:
(WebCore::RenderSVGShape::strokeBoundingBox const):
(WebCore::RenderSVGShape::approximateStrokeBoundingBox const):
(WebCore::RenderSVGShape::calculateApproximateScalingStrokeBoundingBox const):
(WebCore::RenderSVGShape::calculateApproximateNonScalingStrokeBoundingBox const):
(WebCore::RenderSVGShape::calculateApproximateStrokeBoundingBox const):
* Source/WebCore/rendering/svg/RenderSVGShape.h:
* Source/WebCore/rendering/svg/SVGBoundingBoxComputation.cpp:
(WebCore::SVGBoundingBoxComputation::handleShapeOrTextOrInline const):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGContainer.cpp:
(WebCore::LegacyRenderSVGContainer::updateCachedBoundaries):
(WebCore::LegacyRenderSVGContainer::strokeBoundingBox const):
(WebCore::LegacyRenderSVGContainer::repaintRectInLocalCoordinates const):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGContainer.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGPath.cpp:
(WebCore::LegacyRenderSVGPath::updateShapeFromElement):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRoot.cpp:
(WebCore::LegacyRenderSVGRoot::updateCachedBoundaries):
(WebCore::LegacyRenderSVGRoot::strokeBoundingBox const):
(WebCore::LegacyRenderSVGRoot::repaintRectInLocalCoordinates const):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRoot.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.cpp:
(WebCore::LegacyRenderSVGShape::strokeBoundingBox const):
(WebCore::LegacyRenderSVGShape::calculateApproximateScalingStrokeBoundingBox const):
(WebCore::LegacyRenderSVGShape::calculateApproximateNonScalingStrokeBoundingBox const):
(WebCore::LegacyRenderSVGShape::calculateApproximateStrokeBoundingBox const):
(WebCore::LegacyRenderSVGShape::updateRepaintBoundingBox):
(WebCore::LegacyRenderSVGShape::repaintRectInLocalCoordinates const):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.h:

Canonical link: https://commits.webkit.org/269492@main
